### PR TITLE
Feat implement storage cleaner node

### PR DESCRIPTION
**Pull request by Gilbishkosma**
## What does this PR do?

This is a PR for the issue [#227](https://github.com/appwrite/templates/issues/227).
In this i had to create a template to removes all files from the buckets older than X number of days.

## Test Plan
I have tested the function template in Appwrite Cloud,
I created 5buckets and added 10images on each bucket, then deleted the files by executing the template.
Here is the video for it: [demo video](https://drive.google.com/file/d/1fwHGXDygnzK7xovUoKJgdF-uFrEmyNGS/view)

## Related PRs and Issues
This is a PR for the issue [#227](https://github.com/appwrite/templates/issues/227)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes, i have Read the contribution guide.

### CODE REVIEW

**Comment by Gilbishkosma** [comment]
The only concern i have for this template is that it will take time more than 15minutes(this is the maximum execution time for a function) in case the number of files to be deleted is more than 2k. Cause currently there is no method to do bulk delete in appwrite.

File: node/storage-cleaner/package.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "starter-template",
```
Comment (Meldiron) [inline node/storage-cleaner/package.json:2]: Let's name template here too

File: node/storage-cleaner/src/main.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    'APPWRITE_API_KEY',
+  ]);
+
+  const appwrite = new AppwriteService();
+
+  try {
+    await appwrite.cleanAllBuckets();
+  } catch (exception) {
+    error(exception.message);
+  }
```
Comment (Meldiron) [inline node/storage-cleaner/src/main.js:16]: Let's avoid try/catch, so it marks execution as failed if there is an exception.

File: node/storage-cleaner/README.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+| Setting           | Value         |
+| ----------------- | ------------- |
+| Runtime           | Node (18.0)   |
+| Entrypoint        | `src/main.js` |
+| Build Commands    | `npm install` |
+| Permissions       | `any`         |
+| CRON              | `0 1 * * *`   |
+| Timeout (Seconds) | 900           |
+
+## 🔒 Environment Variables
```
Comment (Meldiron) [inline node/storage-cleaner/README.md:26]: Let's introduce `APPWRITE_BUCKET_ID` env variable, and let's only delete files from that one specific bucket.

File: node/storage-cleaner/src/appwrite.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+   * based on the retention period.
+   *
+   * @returns {Promise<void>} A Promise that resolves when all buckets are cleaned.
+   */
+  async cleanAllBuckets() {
+    const buckets = await this.getBuckets();
+    for (const bucket of buckets) {
+      await this.cleanBucket(bucket.$id);
+    }
+  }
```
Comment (Meldiron) [inline node/storage-cleaner/src/appwrite.js:31]: As comented in README, let's only clean one bucket based on ENV variable.

Comment (Meldiron) [reply node/storage-cleaner/src/appwrite.js:31]: We can probably remove the whole `getBuckets` as well.

Comment (Meldiron) [reply node/storage-cleaner/src/appwrite.js:31]: We likely also dont need `defaultQuery` anymore

File: node/storage-cleaner/src/appwrite.js [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+   *
+   * This function retrieves files from the specified bucket, ordered by creation date in ascending order,
+   * and deletes files that are older than the calculated expiry date based on the retention period.
+   *
+   * @param {string} bucketId - The ID of the storage bucket to clean.
+   * @returns {Promise<void>} A Promise that resolves when the bucket is cleaned.
+   */
+  async cleanBucket(bucketId) {
+    const queries = [Query.orderAsc('$createdAt'), Query.limit(100)];
+    const batchPromises = []; // will use batchPromises to send multiple delete request simultaneously
```
Comment (Meldiron) [inline node/storage-cleaner/src/appwrite.js:69]: Let's avoid batching here. It's fine to do one by one. Slow script isn't issue as it's background task.

File: node/storage-cleaner/src/appwrite.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+   * Clean up files from the storage bucket by removing files older than a specified retention period.
+   *
+   * This function retrieves files from the specified bucket, ordered by creation date in ascending order,
+   * and deletes files that are older than the calculated expiry date based on the retention period.
+   *
+   * @param {string} bucketId - The ID of the storage bucket to clean.
+   * @returns {Promise<void>} A Promise that resolves when the bucket is cleaned.
+   */
+  async cleanBucket(bucketId) {
+    const queries = [Query.orderAsc('$createdAt'), Query.limit(100)];
```
Comment (Meldiron) [inline node/storage-cleaner/src/appwrite.js:68]: We should query, something along lines  `Query.smallerThan('$createdAt', getExpiryDate())`

File: node/storage-cleaner/src/appwrite.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    const batchPromises = []; // will use batchPromises to send multiple delete request simultaneously
+    outerLoop: while (true) {
```
Comment (Meldiron) [inline node/storage-cleaner/src/appwrite.js:70]: `while true` is a bit scary. Let's do something along lines:

```js
hasNextPage = true;

do {
  const response = getFIles();
  for(file in response.files) {
   deleteFile(file.$id);
  }
  
  hasNextPage = response.files.length > 0;
} while(hasNextPage);
```

Comment (Meldiron) [reply node/storage-cleaner/src/appwrite.js:70]: here is batching example:

```js
await Promise.all(response.files.map(file) => deleteFile(file.$id));
```

**Review by Meldiron — CHANGES_REQUESTED** [review]
Requested some important changes, let's apply those. Feel free to continue discussion if you are not onboard with my approach.

File: node/storage-cleaner/package.json (hunk 1 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/package.json]: done

File: node/storage-cleaner/README.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
+# Node.js Storage Cleaner Function
```
[the remark this answers is not in the document]
Comment (Gilbishkosma) [reply node/storage-cleaner/README.md]: done

File: node/storage-cleaner/src/main.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import AppwriteService from './appwrite.js';
+import { throwIfMissing } from './util.js';
+
+export default async ({ req, res, log, error }) => {
+  throwIfMissing(process.env, [
+    'APPWRITE_FUNCTION_PROJECT_ID',
```
[the remark this answers is not in the document]
Comment (Gilbishkosma) [reply node/storage-cleaner/src/main.js]: done

File: node/storage-cleaner/src/main.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
+import AppwriteService from './appwrite.js';
+import { throwIfMissing } from './util.js';
+
+export default async ({ req, res, log, error }) => {
+  throwIfMissing(process.env, [
```
[the remark this answers is not in the document]
Comment (Gilbishkosma) [reply node/storage-cleaner/src/main.js:5]: done

File: node/storage-cleaner/src/main.js (hunk 2 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/src/main.js]: done

File: node/storage-cleaner/README.md (hunk 3 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/README.md:26]: done

File: node/storage-cleaner/src/util.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
+/**
+ * Calculate and return the expiry date based on the retention period.
+ * @returns {Date} The calculated expiry date.
+ */
+export function getExpiryDate() {
+  const retentionPeriod = process.env.RETENTION_PERIOD_DAYS ?? 30;
```
[the remark this answers is not in the document]
Comment (Gilbishkosma) [reply node/storage-cleaner/src/util.js]: done

File: node/storage-cleaner/src/appwrite.js (hunk 4 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/src/appwrite.js]: Done, the do while change is good, it made the code smaller.

File: node/storage-cleaner/src/appwrite.js (hunk 6 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/src/appwrite.js]: Done

File: node/storage-cleaner/src/appwrite.js (hunk 7 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/src/appwrite.js]: Done, the do while change is good, it made the code smaller.

**Comment by Gilbishkosma** [comment]
@Meldiron i have pushed the recommended changes

File: node/storage-cleaner/src/appwrite.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+   * Clean up files from the storage bucket by removing files older than a specified retention period.
+   *
+   * @param {string} bucketId - The ID of the storage bucket to clean.
+   * @returns {Promise<void>} A Promise that resolves when the bucket is cleaned.
+   */
+  async cleanBucket(bucketId) {
+    let hasNextPage = true;
+    const queries = [
+      Query.lessThan('$createdAt', getExpiryDate()),
+      Query.limit(100),
```
Comment (gewenyu99) [inline node/storage-cleaner/src/appwrite.js:25]: Can we do smaller batches so we don't risk running out of memory/spam the Storage API unnecessarily? 

@Meldiron thoughts?

File: node/storage-cleaner/src/appwrite.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    ];
+    do {
```
Comment (gewenyu99) [inline node/storage-cleaner/src/appwrite.js:27]: What's the behavior if the bucket is already empty?

**Review by gewenyu99 — CHANGES_REQUESTED** [review]

File: node/storage-cleaner/src/appwrite.js (hunk 12 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/src/appwrite.js:25]: we can change it to 25 instead of 100

Comment (Gilbishkosma) [reply node/storage-cleaner/src/appwrite.js]: I have updated it

File: node/storage-cleaner/src/appwrite.js [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+      const response = await this.storage.listFiles(bucketId, queries);
+      await Promise.all(
+        response.files.map((file) =>
+          this.storage.deleteFile(bucketId, file.$id)
+        )
+      );
+      hasNextPage = response.files.length > 0;
+    } while (hasNextPage);
```
Comment (gewenyu99) [inline node/storage-cleaner/src/appwrite.js:35]: I wonder if we really need the   `  let hasNextPage = true;` var, can we just check it in `response.files.length > 0` ?

I feel like it's readable enough here.

**Review by gewenyu99 — APPROVED** [review]

File: node/storage-cleaner/src/appwrite.js (hunk 14 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/src/appwrite.js:35]: One other method can be to use recursion here, we can call the same cleanBucket again if the response.files.length is greater than 0.

**Review by Meldiron — APPROVED** [review]

File: node/storage-cleaner/src/appwrite.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+      Query.limit(25),
+    ];
+    do {
+      const response = await this.storage.listFiles(bucketId, queries);
+      await Promise.all(
+        response.files.map((file) =>
+          this.storage.deleteFile(bucketId, file.$id)
+        )
+      );
+      hasNextPage = response.files.length > 0;
```
Comment (loks0n) [inline node/storage-cleaner/src/appwrite.js:34]: Can we move this condition inline?

e.g. `} while (response.files.length > 0);`

File: node/storage-cleaner/src/util.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+ * Throws an error if any of the keys are missing from the object
+ * @param {*} obj
+ * @param {string[]} keys
+ * @throws {Error}
+ */
+export function throwIfMissing(obj, keys) {
+  const missing = [];
+  for (let key of keys) {
+    // using obj[key] != 0 for case when the value of key will be 0
+    if (!(key in obj) || (!obj[key] && obj[key] != 0)) {
```
Comment (loks0n) [inline node/storage-cleaner/src/util.js:22]: I think this condition can be simplified:
`!(key in obj && obj[key] !== 0)`

Can we also remove the comment?

File: node/storage-cleaner/src/util.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Calculate and return the expiry date based on the retention period.
+ * @returns {string} The calculated expiry date in ISO format
+ */
+export function getExpiryDate() {
+  const retentionPeriod = +(process.env.RETENTION_PERIOD_DAYS ?? 30);
+  const expiryDate = new Date();
+  expiryDate.setDate(expiryDate.getDate() - retentionPeriod);
+  return expiryDate.toISOString();
+}
```
Comment (loks0n) [inline node/storage-cleaner/src/util.js:10]: How about something like this?
```js
/**
 * Returns a date subtracted by the retention period from the current date.
 * The retention period is fetched from the RETENTION_PERIOD_DAYS environment variable.
 * Defaults to 30 days if the environment variable is not set or invalid.
 * @returns {Date} The calculated expiry date.
 */
export function getExpiryDate() {
  const retentionPeriod = Number(process.env.RETENTION_PERIOD_DAYS ?? 30);
  return new Date(Date.now() - (retentionPeriod * 24 * 60 * 60 * 1000)).toISOString()
}
```

**Review by loks0n — CHANGES_REQUESTED** [review]
Amazing PR! Few small things before we merge

File: node/storage-cleaner/src/appwrite.js (hunk 15 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/src/appwrite.js]: Done

File: node/storage-cleaner/src/util.js (hunk 17 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/src/util.js]: Done

File: node/storage-cleaner/src/util.js (hunk 16 above)
Comment (Gilbishkosma) [reply node/storage-cleaner/src/util.js]: Done

**Comment by Gilbishkosma** [comment]
@loks0n I have pushed the requested changes

**Review by loks0n — APPROVED** [review]

**Comment by loks0n** [comment]
🥳 Merged!

Thanks for your contribution.

**Comment by Gilbishkosma** [comment]
Thanks a lot @loks0n , @Meldiron , @gewenyu99 , @stnguyen90, @Haimantika  😇.
I really enjoyed working on this PR and thanks for the PR watch parties.

**Comment by gewenyu99** [comment]
Hey there! There were a lot of big PRs during this Hacktoberfest, and we wanted to give everyone ample time to collaborate with our engineering team. If you were able to merge your PRs during October, amazing. If it’s still not merged, don’t worry about it either. Either way, we’ve got your Hacktoberfest swag minted and ready to ship. 

Please comment with your Discord username here so we can contact you about your shipping information to deliver your Hacktoberfest swag.

**Comment by Gilbishkosma** [comment]
> Hey there! There were a lot of big PRs during this Hacktoberfest, and we wanted to give everyone ample time to collaborate with our engineering team. If you were able to merge your PRs during October, amazing. If it’s still not merged, don’t worry about it either. Either way, we’ve got your Hacktoberfest swag minted and ready to ship.
> 
> Please comment with your Discord username here so we can contact you about your shipping information to deliver your Hacktoberfest swag.

@gewenyu99 my discord username is: gilbish

**Comment by gewenyu99** [comment]
Be in touch soon!

**Pull request merged**